### PR TITLE
festival encode decode url

### PR DESCRIPTION
**Pull request by yspringfield**
fixed: #7065

### CODE REVIEW

File: apps/festival/festival/src/app/marketplace/title/list/list.component.ts [hunk 1]
```diff
@@ -74,11 +77,45 @@ export class ListComponent implements OnInit, OnDestroy {
...
+    });
+    this.subs.push(sub);
   }
 
+  ngAfterViewInit(): void {
+    const {
+      storeStatus, genres, originCountries, productionStatus, sellers, socialGoals, languages,
+      query, minBudget, ...rest
+    }: MovieSearch = decodeUrl(this.route);
+    console.log({rest, query,
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/marketplace/title/list/list.component.ts:89]: remove console.log

File: apps/festival/festival/src/app/marketplace/title/list/list.component.ts [hunk 2]
```diff
@@ -74,11 +77,45 @@ export class ListComponent implements OnInit, OnDestroy {
...
+      storeStatus, genres, originCountries, productionStatus, sellers, socialGoals, languages,
+    })
+    const sub = this.searchForm.valueChanges.pipe(
+      throttleTime(1000),
+    ).subscribe(formState => {
+      encodeUrl<MovieSearch>(this.router, this.route, formState);
+    });
+
+    // patching the formlists and formEntities
+    this.searchForm.query.patchValue(query);
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/marketplace/title/list/list.component.ts:99]: Prefer : 
```typescript
const search: MovieSearch = decodeUrl(this.route);
this.searchForm.patchAllValue(search);
const sub = this.searchForm.valueChanges.pipe(
  throttleTime(1000)
).subscribe(value => encodeUrl(this.router, this.route, value));
this.subs.push(sub);
```

File: apps/festival/festival/src/app/marketplace/title/list/list.component.ts [hunk 3]
```diff
@@ -89,6 +126,8 @@ export class ListComponent implements OnInit, OnDestroy {
   }
 
   ngOnDestroy() {
-    this.sub.unsubscribe();
+    this.subs.forEach(element => {
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/marketplace/title/list/list.component.ts:129]: inline tis : `this.subs.forEach(sub => sub.unsubcribe())`

File: libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts [hunk 4]
```diff
@@ -51,6 +51,7 @@ export class LanguageFilterComponent implements OnInit, OnDestroy {
 
     /** Updates selectedLanguages and versions FormList when reset is called on languagesFilterForm */
     this.formSub = this.languagesFilterForm.valueChanges.subscribe((res: Array<any>) => {
+      console.log({res})
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts:54]: remove console.log

File: apps/festival/festival/src/app/marketplace/title/list/list.component.ts (hunk 2 above)
Comment (yspringfield) [reply apps/festival/festival/src/app/marketplace/title/list/list.component.ts]: Calling patchAllValue on the search form doesn't actually updates the underlying formEntities and FormLists.

File: apps/festival/festival/src/app/marketplace/title/list/list.component.html [hunk 5]
```diff
@@ -47,11 +47,12 @@ <h1 class="mat-display-2 page-title">On the market today</h1>
     </ng-template>
 
     <ng-template filter label="Language & Version" [form]="searchForm.languages">
-      <title-language-filter [languagesFilterForm]="searchForm.languages"></title-language-filter>
+      <title-language-filter [selectedLanguages$]="selectedLanguages$" [languagesFilterForm]="searchForm.languages"></title-language-filter>
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/marketplace/title/list/list.component.html:50]: you should not pass an observable as input. Also the information is already in `searchForm.languages` so there should be a clean way to get it work.

File: libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts [hunk 6]
```diff
@@ -62,11 +66,28 @@ export class LanguageFilterComponent implements OnInit, OnDestroy {
...
       }
     })
+    if (this.selectedLanguages$) {
+      const sub2 = this.selectedLanguages$.subscribe(data => this.hardReset(data))
+      this.subs.push(sub2)
+    }
+    this.subs.push(sub, sub1);
+  }
+
+  hardReset(data?: LanguagesSearch) {
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts:76]: component shouldn't be the one responsible to update the content of form like that

**Review by GrandSchtroumpf — COMMENTED** [review]
This is a dirty way to solve the issue. Please remove everything related to `title-language-filter` update and patch the form without destructuring the data from the url

**Comment by RemcoSimonides** [comment]
Hee @yspringfield , are you still working on this PR? would be super useful to have for solving a task I'm working on in issue #7172

File: apps/festival/festival/src/app/marketplace/title/list/list.component.ts [hunk 7]
```diff
@@ -74,9 +78,35 @@ export class ListComponent implements OnInit, OnDestroy {
...
+    this.subs.push(sub);
   }
 
+  ngAfterViewInit(): void {
+    const {
+      storeStatus, genres, originCountries, productionStatus,
+      sellers, socialGoals, languages, query, minBudget
+    }: MovieSearch = decodeUrl(this.route);
+
+    // patching the formlists and formEntities
```
[the remark this answers is not in the document]
Comment (yspringfield) [reply apps/festival/festival/src/app/marketplace/title/list/list.component.ts:92]: Using patchAllValues doesn't updates the constituent formControls.

[the remark this answers is not in the document]
Comment (yspringfield) [reply apps/festival/festival/src/app/marketplace/title/list/list.component.ts:92]: calling patchAllValues doesn't update the constituent formControls and formGroups.

File: libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts [hunk 8]
```diff
@@ -1,70 +1,26 @@
...
 
-  @Input() languagesFilterForm: FormEntity<LanguageVersionControl>; // FormGroup of FormArray
+  @Input() form: LanguageVersionControl;
 
-  /** list of selected language (chips), they can later be added in *'original'*, *'subtitle'*, etc... */
-  public selectedLanguages = FormList.factory<GetKeys<'languages'>>([]);
+  public items = staticModel['movieLanguageTypes'];
 
-  public versions = FormList.factory<GetKeys<'movieLanguageTypes'>>(['original', 'dubbed', 'subtitle'], el => new FormControl(el))
+  private subs: Subscription[] = [];
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts:18]: looks like you're not using that.

File: libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts [hunk 9]
```diff
@@ -1,70 +1,26 @@
...
-        this.languagesFilterForm.markAsDirty();
-      }
-    })
-  }
-
-  ngOnDestroy() {
-    this.sub.unsubscribe();
-    this.formSub.unsubscribe();
-  }
+  constructor(private cdr: ChangeDetectorRef) { }
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts:23]: looks like you're not using that.

File: libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts [hunk 10]
```diff
@@ -1,70 +1,26 @@
...
 
+  public keepOrder = () => 1;
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/filters/languages/language-filter.component.ts:25]: should be before the constructor

File: libs/movie/src/lib/movie/form/search.form.ts [hunk 11]
```diff
@@ -150,10 +162,11 @@ export class MovieSearchForm extends FormEntity<MovieSearchControl> {
       this.storeStatus.value?.length === 0 &&
       this.genres.value?.length === 0 &&
       this.originCountries.value?.length === 0 &&
-      this.languages.value?.original.length === 0 &&
-      this.languages.value?.dubbed.length === 0 &&
-      this.languages.value?.subtitle.length === 0 &&
-      this.languages.value?.caption.length === 0 &&
+      (!(this.languages.value?.languages?.length === 0) ||
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/search.form.ts:165]: Create an intermediary variable :
```typescript
isEmpty() {
  const { caption, dubbed, original } = this.languages.value?.versions;
  const emptyVersions = !caption && !dubbed && !original;
  const emptyLanguages = !this.languages.value?.languages?.length || emptyVersions;
  return {
    .. &&
    emptyLanguages &&
    ...
}
```

File: libs/utils/src/lib/form/forms/list.form.ts [hunk 12]
```diff
@@ -148,4 +148,11 @@ export class FormList<T, Control extends AbstractControl = any> extends FormArra
   ) {
     super.patchValue(value, options)
   }
+
+  hardReset(values: GetPartial<T>[]) {
+    this.clear()
+    values.forEach(value => {
+      this.add(value)
+    })
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/form/forms/list.form.ts:156]: inline it 
```suggestion
    values.forEach(value => this.add(value));
```

**Review by GrandSchtroumpf — APPROVED** [review]

**Pull request merged**